cognitive/src/main/scala/com/microsoft/ml/spark/cognitive/TextTranslator.scala

```diff
@@ -0,0 +1,406 @@
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.{ArrayType, DataType, StructType}
+import spray.json.DefaultJsonProtocol._
+import spray.json._
+
+import java.net.URI
+
+trait HasSubscriptionRegion extends HasServiceParams {
+  val subscriptionRegion = new ServiceParam[String](
+    this, "subscriptionRegion", "the API region to use")
+
+  def getSubscriptionRegion: String = getScalarParam(subscriptionRegion)
+
+  def setSubscriptionRegion(v: String): this.type = setScalarParam(subscriptionRegion, v)
+
+  def getSubscriptionRegionCol: String = getVectorParam(subscriptionRegion)
+
+  def setSubscriptionRegionCol(v: String): this.type = setVectorParam(subscriptionRegion, v)
+
+}
+
+trait HasTextInput extends HasServiceParams {
+  val text = new ServiceParam[Seq[String]](
+    this, "text", "the string to translate")
+
+  def getText: Seq[String] = getScalarParam(text)
+
+  def setText(v: Seq[String]): this.type = setScalarParam(text, v)
+
+  def getTextCol: String = getVectorParam(text)
+
+  def setTextCol(v: String): this.type = setVectorParam(text, v)
+
+}
+
+trait HasFromLanguage extends HasServiceParams {
+  val fromLanguage = new ServiceParam[String](this, "fromLanguage", "Specifies the language of the input text." +
+    " The source language must be one of the supported languages included in the dictionary scope.",
+    isRequired = true, isURLParam = true)
+
+  def setFromLanguage(v: String): this.type = setScalarParam(fromLanguage, v)
+
+  def setFromLanguageCol(v: String): this.type = setVectorParam(fromLanguage, v)
+
+  def getFromLanguage: String = getScalarParam(fromLanguage)
+
+  def getFromLanguageCol: String = getVectorParam(fromLanguage)
+}
+
+trait HasToLanguage extends HasServiceParams {
+  val toLanguage = new ServiceParam[String](this, "toLanguage", "Specifies the language of the output text." +
+    " The target language must be one of the supported languages included in the dictionary scope.",
+    isRequired = true, isURLParam = true)
+
+  def setToLanguage(v: String): this.type = setScalarParam(toLanguage, v)
+
+  def setToLanguageCol(v: String): this.type = setVectorParam(toLanguage, v)
+
+  def getToLanguage: String = getScalarParam(toLanguage)
+
+  def getToLanguageCol: String = getVectorParam(toLanguage)
+}
+
+trait TextAsOnlyEntity extends HasTextInput with HasCognitiveServiceInput {
+
+  override protected def prepareEntity: Row => Option[AbstractHttpEntity] = {
+    r =>
+      Some(new StringEntity(
+        getValueOpt(r, text)
+          .map(x => x.map(y => Map("Text" -> y))).toJson.compactPrint, ContentType.APPLICATION_JSON))
+  }
+}
+
+abstract class TextTranslatorBase(override val uid: String) extends CognitiveServicesBase(uid)
+  with HasInternalJsonOutputParser with HasCognitiveServiceInput with HasSubscriptionRegion
+  with HasSetLocation {
+
+  protected val subscriptionRegionHeaderName = "Ocp-Apim-Subscription-Region"
```

Comment:
this is so strange i cant believe they make you specify this in a header lol

Reply:
Yes it's quite strange... the document says it's optional for global translator resource, but if we don't add it into header the response will be '{"error":{"code":401000,"message":"The request is not authorized because credentials are missing or invalid."}}'

Reply:
Arent cognitive services so great!